### PR TITLE
content: Change `UserMention` to render user mentions dynamically

### LINKED ISSUE #1258: Handle user renames in rendering @-mentions
Currently we render @-mentions using the user's name as recorded in the HTML. If the mentioned user has edited their name since the message was sent (or was last edited), this will be their old name. Instead, we should show the mentioned user's current name.

To do this we'll use the mentioned user's ID, which is also encoded in the HTML of the @-mention. Then we can look that up in our users data. If the user is unknown, we can fall back to the name that's in the HTML.

This behavior is specified in docs just written today:
https://zulip.com/api/message-formatting#mentions
(or maybe [this link](https://zulip.com/api/message-formatting#mentions-and-silent-mentions); added in https://github.com/zulip/zulip/commit/3aa267167934086d181f1c9bae4c57c5df7e3756), but I believe it's been the web app's behavior for a long time.

### Related issues
* #1047 
* #1259

### PR DESCRIPTION
**Pull request by Ruhdee**
Fixes zulip#1258

| Before | After |
|--------|-------|
|Screenshot_20260117-131618 Zulip| Screenshot_20260117-132059 Zulip|
|Screenshot_20260117-131627 Zulip|Screenshot_20260117-132050 Zulip|
| User mention shows old name "Old Name" | User mention shows current name "New Name" from store |

#### Model:
Added userId, isSilent and updated parser.

#### Widget:
Modified widget to fetch username from store and handle isSilent flag.

#### Tests:
Updated parser tests and added new widget tests.

### Doubts:

1. Running flutter analyze gives error on code from main branch as mentioned in zulip#2082

UPDATE: Removing the error causing annotations fixes the CI testing issue. However, since this code change is not related to issue zulip#1258, please advise on how I should proceed.

2. Old widget tests for UserMention don't pass store. So I've added defensive check for store in UserMention widget even though store will probably never be null in production. Should I keep the code as it is or should I change all old tests for UserMention?

### CODE REVIEW

**Comment by gnprice** [comment]
Thanks! This looks quite promising.

Here's some high-level comments. Once you address these, the PR will be ready for a more detailed maintainer review.

First, to your questions:

> Running flutter analyze gives error on code from main branch

Yeah, including a fix for that is fine as long as it's a separate commit, which you've done.

> Old widget tests for UserMention don't pass store. So I've added defensive check for store in UserMention widget even though store will probably never be null in production.

Yeah, we don't want to add that sort of logic. It makes the code more complicated to read and think through, and it makes the behavior in tests diverge from the behavior in the real app.

Instead, any existing tests that start needing a store widget should add one. Lots of our other existing tests provide a store widget, because lots of our other widget code requires one, so it should be possible to do this with very little added code. This change can happen as its own separate prep commit, before the change that makes it necessary.

---

One of your later commits changes some existing tests so that they pass. Those changes should instead happen in the same commits that make them necessary. In general, if you check out any individual commit and run `tools/check` there, it should pass.

Another later commit adds tests for an earlier commit. Instead, those tests should be added in the same commit as the code they test.

**Comment by Ruhdee** [comment]
Changes made:
1. Rebased onto main. This fixed the `@override` annotations error.
> Instead, any existing tests that start needing a store widget should add one. Lots of our other existing tests provide a store widget, because lots of our other widget code requires one, so it should be possible to do this with very little added code. This change can happen as its own separate prep commit
2. Added preparatory tests commit according to above
3. Squashed tests commits to where model or widget changes were made

Please review changes.

Small question: Is it ok to tag maintainers, like so: @gnprice, when feedback is addressed to speed up the review process? The contributing guide doesn't specifically advise for or against it in the "How to help move the review process forward" section.

**Comment by gnprice** [comment]
Cool, thanks for the revision. This looks ready for a more detailed review.

> Is it ok to tag maintainers, like so: @gnprice, when feedback is addressed to speed up the review process?

It's not really harmful to tag maintainers on a PR thread, but doesn't have much effect. I already watch this repo and so get notifications for every update to every PR.

More generally on GitHub, anyone who's already participating in the thread can be expected to be subscribed and so getting notifications for every update, with or without an @-mention. That means that GitHub @-mentions are mainly useful for pulling someone into a thread for the first time.

The main thing to do when you've addressed feedback is just to post a comment clearly saying you've done so. That way when a maintainer looks at it, it's clear that it's ready for the next review and not still waiting for a revision from you.

**Comment by Ruhdee** [comment]
@rajveermalviya please review. Since this is my first PR, I'm not allowed to work on other issues until this one gets merged.

File: lib/model/content.dart [hunk 1]
```diff
@@ -1120,11 +1129,15 @@ class _ZulipInlineContentParser {
       return null;
     }
 
+    final userIdString = element.attributes['data-user-id'];
+    final userId = (userIdString == null || userIdString == '*') ? null
+      : int.tryParse(userIdString);
```
Comment (rajveermalviya) [inline lib/model/content.dart]: nit: simplify
```suggestion
    final userId = switch (element.attributes['data-user-id']) {
      null || '*' => null,
      final userIdString => int.tryParse(userIdString),
    };
```

Comment (rajveermalviya) [reply lib/model/content.dart]: Also would be good to add a small comment for the first case, explaining why we are returning null here.

File: lib/model/content.dart [hunk 2]
```diff
@@ -969,15 +969,24 @@ class UserMentionNode extends InlineContainerNode {
   const UserMentionNode({
     super.debugHtmlNode,
     required super.nodes,
+    required this.userId,
   });
 
+  final int? userId;
```
Comment (rajveermalviya) [inline lib/model/content.dart:1088]: Let's have a dartdoc here explaining what this value is, and when it will be null. (Then we can remove that information from the commit message.)

File: test/widgets/content_test.dart
Comment (rajveermalviya) [inline test/widgets/content_test.dart:1]:
```
tests: Add store widget to some existing tests
…
```

The correct commit header prefix to use here would be `content test:`.

You can lookup what header prefix (or general commit message style) previous commits used for a specific file using this git command: `git log test/model/content_test.dart`

Comment (rajveermalviya) [reply test/widgets/content_test.dart:1]: Additionally, looks like this commit is basically just a refactor and doesn't affect the actual logic, right? If so this commit can also be marked as a "non-functional change". To do that just append `[nfc]`, so the header prefix then will become `content test [nfc]:`.

File: lib/widgets/content.dart [hunk 3]
```diff
@@ -1223,7 +1223,13 @@ class UserMention extends StatelessWidget {
 
   @override
   Widget build(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
     final contentTheme = ContentTheme.of(context);
+    final user = node.userId != null ? store.getUser(node.userId!) : null;
+    final displayText = user != null
+      ? (node.isSilent ? user.fullName : '@${user.fullName}') 
+      : null;
+    final nodes = displayText != null ? [TextNode(displayText)] : node.nodes;
```
Comment (rajveermalviya) [inline lib/widgets/content.dart]: nit: simplify, and avoid using nested `?:` statements.

```suggestion
    var nodes = node.nodes;
    if (node.userId case final userId?) {
      final user = store.getUser(userId);
      if (user case User(:final fullName)) {
        nodes = [TextNode(node.isSilent ? fullName : '@$fullName')];
      }
    }
```

The above snippet uses Dart's pattern matching syntax, you can learn more about it by reading here: https://dart.dev/language/patterns

**Review by rajveermalviya — CHANGES_REQUESTED** [review]
Thanks for working on this @Ruhdee, great to have this fixed. Generally all looks good, just some small comments below.

**Comment by Ruhdee** [comment]
@rajveermalviya I have made all the required changes. Thank you for the helpful tips and doc links! Please review.

Small question: Should I add dartdoc for 'isSilent' field? Or is its purpose sufficiently self-explanatory?

**Review by rajveermalviya — APPROVED** [review]
Thanks! LGTM. Moving over to Greg's review.

> Small question: Should I add dartdoc for 'isSilent' field? Or is its purpose sufficiently self-explanatory?

IMO it would be fine leaving that out, `int? userId` needed one because it was unclear when it would be null.

File: lib/model/content.dart [hunk 4]
```diff
@@ -1318,11 +1138,21 @@ class _ZulipInlineContentParser {
...
+    };
+
     // TODO assert UserMentionNode can't contain LinkNode;
     //   either a debug-mode check, or perhaps we can make expectations much
     //   tighter on a UserMentionNode's contents overall.
     final nodes = parseInlineContentList(element.nodes);
-    return UserMentionNode(nodes: nodes, debugHtmlNode: debugHtmlNode);
+    return UserMentionNode(
+      nodes: nodes,
+      debugHtmlNode: debugHtmlNode,
```
Comment (gnprice) [inline lib/model/content.dart]: nit: keep debugHtmlNode as the last argument (similar to other node constructor calls here)

File: lib/model/content.dart [hunk 5]
```diff
@@ -1077,15 +969,31 @@ class UserMentionNode extends InlineContainerNode {
...
     super.debugHtmlNode,
     required super.nodes,
+    required this.userId,
+    required this.isSilent,
   });
 
+  /// The ID of the user being mentioned.
+  ///
+  /// This is null for wildcard mentions, user group mentions, topic mentions
+  /// or when the user ID is unavailable in the HTML (e.g., legacy mentions).
```
Comment (gnprice) [inline lib/model/content.dart]: nit: a topic mention is a type of wildcard mention
```suggestion
  /// This is null for wildcard mentions, user group mentions,
  /// or when the user ID is unavailable in the HTML (e.g., legacy mentions).
```

See docs here: https://zulip.com/api/message-formatting#mentions-and-silent-mentions

Same remark applies to a comment below.

File: test/model/content_test.dart [hunk 6]
```diff
@@ -111,105 +111,105 @@ class ContentExample {
     "@**Greg Price**",
     expectedText: '@Greg Price',
     '<p><span class="user-mention" data-user-id="2187">@Greg Price</span></p>',
-    const UserMentionNode(nodes: [TextNode('@Greg Price')]));
+    const UserMentionNode(nodes: [TextNode('@Greg Price')], userId: 2187, isSilent: false));
```
Comment (gnprice) [inline test/model/content_test.dart:114]: nit: no need to mention this bit in the commit messages:

    Also changed 'UserMentionNode' calls in content model tests.

It's a necessary part of making the commit work, and is unsurprising given you're adding the field.

File: lib/model/content.dart [hunk 7]
```diff
@@ -1077,15 +969,31 @@ class UserMentionNode extends InlineContainerNode {
...
+  final int? userId;
+
+  final bool isSilent; // TODO(#647)
```
Comment (gnprice) [inline lib/model/content.dart:1090]: nit: in commit message:

     false for regular mentions. `isSilent` is necessary for zulip#647.

Leave out the "zulip" prefix; just "#647".

Can also make the relationship of that to the current work clearer: "The `isSilent` field will be necessary later for #647." Otherwise it sounds like this might be describing the motivation for making the change now (since after all that's part of the primary subject matter of a commit message: why make this change, and why is it OK to make this change).

File: lib/model/content.dart [hunk 8]
```diff
@@ -1077,15 +969,31 @@ class UserMentionNode extends InlineContainerNode {
...
+
   // For the legacy design, we don't need this information in code; instead,
   // the inner text already shows how to communicate it to the user
   // (e.g., silent mentions' text lacks a leading "@"),
   // and we show that text in the same style for all types of @-mention.
```
Comment (gnprice) [inline lib/model/content.dart:1094]: This comment no longer matches what it's attached to, now that `isSilent` is uncommented and moved above. This fix would be good enough:

```suggestion
// the inner text already shows how to communicate it to the user,
// and we show that text in the same style for all types of @-mention.
```

File: test/widgets/content_test.dart [hunk 9]
```diff
@@ -171,9 +170,10 @@ void main() {
   /// check that the content has actually rendered.  For examples where there's
   /// no suitable value for [ContentExample.expectedText], use [prepareContent]
   /// and write an appropriate content-has-rendered check directly.
-  void testContentSmoke(ContentExample example) {
+  void testContentSmoke(ContentExample example, {bool wrapWithPerAccountStoreWidget = false}) {
     testWidgets('smoke: ${example.description}', (tester) async {
-      await prepareContent(tester, plainContent(example.html));
+      await prepareContent(tester, plainContent(example.html),
+        wrapWithPerAccountStoreWidget: wrapWithPerAccountStoreWidget);
```
Comment (gnprice) [inline test/widgets/content_test.dart:177]: nit: this commit message is inaccurate:

    content test [nfc]: Add store widget to some existing tests

Because the parameter defaults to false and this commit doesn't cause tests to start passing it as true, it doesn't actually change whether the store widget is present in any given test.

The commit is fine, just needs an accurate commit message.

File: test/widgets/content_test.dart [hunk 10]
```diff
@@ -1046,6 +839,44 @@ void main() {
...
+      }) async {
+        final initialSnapshot = users != null
+          ? eg.initialSnapshot(realmUsers: users)
+          : eg.initialSnapshot();
+        await prepareContent(tester,
+          wrapWithPerAccountStoreWidget: true,
+          initialSnapshot: initialSnapshot,
+          plainContent(html));
+      }
+      testWidgets('resolves current user name from store', (tester) async {
```
Comment (gnprice) [inline test/widgets/content_test.dart:1082]: nit: these are both multi-line, so give them separate stanzas by separating with a blank line:

```suggestion
      }

      testWidgets('resolves current user name from store', (tester) async {
```

Comment (gnprice) [reply test/widgets/content_test.dart:1082]: (similarly for between tests below)

File: test/widgets/content_test.dart [hunk 11]
```diff
@@ -1046,6 +839,44 @@ void main() {
...
+
+    group('dynamic name resolution', () {
+      Future<void> prepare({
+        required WidgetTester tester,
+        required String html,
+        List<User>? users,
+      }) async {
+        final initialSnapshot = users != null
+          ? eg.initialSnapshot(realmUsers: users)
+          : eg.initialSnapshot();
```
Comment (gnprice) [inline test/widgets/content_test.dart]: I believe this simpler version is equivalent:
```suggestion
        final initialSnapshot = eg.initialSnapshot(realmUsers: users);
```

File: test/widgets/content_test.dart [hunk 12]
```diff
@@ -1046,6 +839,44 @@ void main() {
...
+      }
+      testWidgets('resolves current user name from store', (tester) async {
+        await prepare(
+          tester: tester,
+          html: '<p><span class="user-mention" data-user-id="123">@Old Name</span></p>',
+          users: [eg.selfUser, eg.user(userId: 123, fullName: 'New Name')]);
+        check(find.text('@New Name')).findsOne();
+        check(find.text('@Old Name')).findsNothing();
+      });
+      testWidgets('falls back to original text when user not found', (tester) async {
```
Comment (gnprice) [inline test/widgets/content_test.dart:1091]: Let's also have a test that exercises the case where `userId` is null.

**Review by gnprice — COMMENTED** [review]
Thanks @Ruhdee, and thanks @rajveermalviya for the previous reviews! This is quite clean. Comments below, all small.

File: test/widgets/content_test.dart (hunk 12 above)
Comment (Ruhdee) [reply test/widgets/content_test.dart:1091]: This case is already covered in the regular smoke tests. Still, I will add a test to mention the behaviour explicitly.

**Comment by Ruhdee** [comment]
@gnprice I have addressed the feedback. PTAL!

File: lib/model/content.dart (hunk 7 above)
Comment (Ruhdee) [reply lib/model/content.dart:1090]: In my local branch, I have removed the 'zulip' prefix (does not show 'zulip' in git log). Yet, when I push it to Github, 'zulip' prefix reappears. 
Screenshot 2026-01-30 180229

Comment (gnprice) [reply lib/model/content.dart:1090]: Yeah, GitHub sometimes shows that in its page for a commit. I think the idea has something to do with the fact that GitHub's model puts your PR branch in a full-blown "fork" of this repo, which in principle could have its own issues and PRs too. So the `zulip` part is to disambiguate that it's talking about issue 647 of this repo (in the `zulip` organization), rather than a hypothetical `Ruhdee#647` which would be an issue or PR 647 in your "fork"'s hypothetical issue tracker.

Anyway the main thing from my perspective is to make it correct in Git. That's the main way I read the commit messages, and the main way anyone studying the codebase is going to do so, because `git log` is a lot [more powerful](https://github.com/zulip/zulip-mobile/blob/main/docs/howto/git.md#git-log-secret) and a lot faster than the GitHub website.

File: test/widgets/content_test.dart (hunk 12 above)
Comment (gnprice) [reply test/widgets/content_test.dart:1091]: Ah, that makes sense. Yeah, good to make it explicit anyway — otherwise it isn't obvious to the reader that that case is covered.

**Comment by gnprice** [comment]
Thanks for the revision! All looks good now, with two nits:

    content: Change `UserMention` to render user mentions dynamically
    
     Look up current user name from PerAccountStore when rendering
     mentions, while respecting "isSilent" flag. This ensures mentions
     display correctly even if the user has changed their name since the
     message was sent. Falls back to original HTML text when user not
     found. Changed old user mention tests to pass store widget.
     Added 'dynamic name resolution' tests group.
    
     Fixes #1258

This commit-message body is oddly indented by a space.

The "Fixes" mention doesn't quite match our style: we say either "Fixes #1258.", with period, at the beginning, or "Fixes: #1258", with colon, at the end. (Usually we prefer the style at the beginning, which sets context for the rest of the commit message. The style at the end, with colon, is traditional for adding a variety of other metadata to a commit message.)

I'll fix those and merge.
